From d71b739fa3762c8d2483d4ca8c0508a05e723484 Mon Sep 17 00:00:00 2001
From: Hector Cao <hector.cao@canonical.com>
Date: Wed, 29 Apr 2026 13:39:50 +0200
Subject: [PATCH] module: fix early stop for load module function

Commit c551fb0b53d regressed module-loading behavior.
Historically, when loading from one module directory failed,
QEMU continued searching the remaining directories. That commit
made QEMU stop if a matching file was found but could not be
loaded (for example, due to a build mismatch).

As a result, QEMU can fail to load the correct module, typically
when the build option module-upgrades is enabled and at upgrades
multiple module directories exist.

In addition to the regression on the behavior, this commit introduces
also a leak in the function module_load_dso() that does not remove
the entries from the list dso_init_list when the the module load
fails for build mismatch.

This commit has 2 folds:
 - restores the correct legacy behavior + set rv to -1 in case
   of load failure for one found module.
 - fix the memory leak in module_load_dso()

Fixes: c551fb0b53d ("module: add Error arguments to module_load and module_load_qom")
Signed-off-by: Hector Cao <hector.cao@canonical.com>
---
 util/module.c | 51 ++++++++++++++++++++++++++++++++++-----------------
 1 file changed, 34 insertions(+), 17 deletions(-)

Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+bug/2146445
Forwarded: https://lists.gnu.org/archive/html/qemu-devel/2026-04/msg06377.html

diff --git a/util/module.c b/util/module.c
index 09dc43f51eb..435b34e354f 100644
--- a/util/module.c
+++ b/util/module.c
@@ -184,6 +184,11 @@ static bool module_load_dso(const char *fname, bool export_symbols,
                 "Only modules from the same build can be loaded.\n");
         }
         g_module_close(g_module);
+        /* Clean up any dso init entries added during DSO initialization */
+        QTAILQ_FOREACH_SAFE(e, &dso_init_list, node, next) {
+            QTAILQ_REMOVE(&dso_init_list, e, node);
+            g_free(e);
+        }
         return false;
     }
 
@@ -279,29 +284,41 @@ int module_load(const char *prefix, const char *name, Error **errp)
         }
     }
 
-    for (i = 0; i < n_dirs; i++) {
+    /*
+     * Try to load module in given folders until we load it successfully or no
+     * more to look.
+     * If no module can be loaded successfully, errp is set to the last encountered error
+     */
+    rv = 0;
+    for (i = 0; (i < n_dirs) && (rv != 1); i++) {
         char *fname = g_strdup_printf("%s/%s%s",
                                       dirs[i], module_name, CONFIG_HOST_DSOSUF);
         int ret = access(fname, F_OK);
-        if (ret != 0 && (errno == ENOENT || errno == ENOTDIR)) {
-            /*
-             * if we don't find the module in this dir, try the next one.
-             * If we don't find it in any dir, that can be fine too: user
-             * did not install the module. We will return 0 in this case
-             * with no error set.
-             */
-            g_free(fname);
-            continue;
-        } else if (ret != 0) {
-            /* most common is EACCES here */
-            error_setg_errno(errp, errno, "error trying to access %s", fname);
-        } else if (module_load_dso(fname, export_symbols, errp)) {
-            rv = 1; /* module successfully loaded */
+        if (ret == 0) {
+            if (module_load_dso(fname, export_symbols, errp)) {
+                /* module successfully loaded */
+                rv = 1;
+            } else {
+                /* fail to load a module */
+                rv = -1;
+            }
+        } else {
+            /* The file exists but cannot be accessed -> report an error and continue */
+            if ((errno != ENOENT) && (errno != ENOTDIR)) {
+                /* most common is EACCES here */
+                error_setg_errno(errp, errno, "error trying to access %s", fname);
+                rv = -1;
+            }
         }
+
         g_free(fname);
-        goto out;
+
+        /* If we have more to look at, report the current error */
+        if ((i < n_dirs - 1) && (rv != 1) && errp && *errp) {
+            error_report_err(*errp);
+            *errp = NULL;
+        }
     }
-    rv = 0; /* module not found */
 
 out:
     if (rv <= 0) {
-- 
2.43.0

